File: backup/demo/cpp/dep_parser.py
#!/usr/bin/env python3

import sys
import json
import re
from pathlib import Path


def parse_makefile_deps(dep_file_path, output_file_path):
    try:
        with open(dep_file_path, 'r') as f:
            content = f.read()

        content = re.sub(r'\\\s*\n\s*', ' ', content)
        match = re.search(r'.*?\s*:\s*(.+)', content)

        if not match:
            print(f"WARNING: cann't parse {dep_file_path}", file=sys.stderr)
            deps = []
        else:
            deps_str = match.group(1).strip()
            deps = [dep.strip() for dep in deps_str.split() if dep.strip()]

        target_file = output_file_path.replace('.dep.json', '.o')
        result = [{
            "file": target_file,
            "deps": deps
        }]

        with open(output_file_path, 'w') as f:
            json.dump(result, f, indent=2)

        print(f"Generated {output_file_path}")
        return True

    except Exception as e:
        print(f"ERROR: parse {dep_file_path} failure: {e}", file=sys.stderr)
        return False


def main():
    if len(sys.argv) != 3:
        print("usage: dep_parser.py <input.d> <output.dep.json>", file=sys.stderr)
        print("example: dep_parser.py build/main.d build/main.dep.json", file=sys.stderr)
        sys.exit(1)

    input_file = sys.argv[1]
    output_file = sys.argv[2]
    if not Path(input_file).exists():
        print(f"ERROR: input file {input_file} not exists", file=sys.stderr)
        sys.exit(1)

    success = parse_makefile_deps(input_file, output_file)
    sys.exit(0 if success else 1)


if __name__ == "__main__":
    main()

File: backup/demo/cpp/dep_parser.cpp
#include <filesystem>
#include <fstream>
#include <iostream>
#include <regex>
#include <sstream>
#include <string>
#include <vector>


bool parseMakefileDeps(const std::string &depFilePath, const std::string &outputFilePath)
{
    std::ifstream file(depFilePath);
    if (!file.is_open()) {
        std::cerr << "ERROR: cann't open file " << depFilePath << std::endl;
        return false;
    }

    std::string content((std::istreambuf_iterator<char>(file)),
                        std::istreambuf_iterator<char>());
    file.close();

    std::regex continuationRegex(R"(\\\s*\n\s*)");
    content = std::regex_replace(content, continuationRegex, " ");
    std::regex depRegex(R"(.*?\s*:\s*(.+))");
    std::smatch match;
    std::vector<std::string> deps;
    if (std::regex_search(content, match, depRegex)) {
        std::string depsStr = match[1].str();
        std::istringstream iss(depsStr);
        std::string dep;
        while (iss >> dep) {
            if (!dep.empty()) {
                deps.push_back(dep);
            }
        }
    }

    std::string targetFile = outputFilePath;
    size_t pos = targetFile.find(".dep.json");
    if (pos != std::string::npos) {
        targetFile.replace(pos, 9, ".o");
    }

    std::ofstream outFile(outputFilePath);
    if (!outFile.is_open()) {
        std::cerr << "ERROR: cann't create output file " << outputFilePath << std::endl;
        return false;
    }

    outFile << "[\n  {\n    \"file\": \"" << targetFile << "\",\n    \"deps\": [";
    for (size_t i = 0; i < deps.size(); ++i) {
        if (i > 0)
            outFile << ",";
        outFile << "\n      \"" << deps[i] << "\"";
    }

    outFile << "\n    ]\n  }\n]\n";
    outFile.close();

    std::cout << "Generated " << outputFilePath << std::endl;
    return true;
}

int main(int argc, char *argv[])
{
    if (argc != 3) {
        std::cerr << "usage: " << argv[0] << " <input.d> <output.dep.json>"
            << std::endl;
        std::cerr << "example: " << argv[0] << " build/main.d build/main.dep.json"
            << std::endl;
        return 1;
    }

    std::string inputFile = argv[1];
    std::string outputFile = argv[2];

    if (!std::filesystem::exists(inputFile)) {
        std::cerr << "ERROR: input file " << inputFile << " not exists" << std::endl;
        return 1;
    }

    return parseMakefileDeps(inputFile, outputFile) ? 0 : 1;
}
